packages/@lwc/engine-dom/src/patches/global-registry.ts

```diff
@@ -128,42 +130,58 @@ function createPivotingClass(tagName: string, registeredDefinition: Definition)
     return PivotCtor;
 }
 
-function getObservedAttributesOffset(
+function getNewObservedAttributes(
     registeredDefinition: Definition,
-    instancedDefinition: Definition
+    instanceDefinition: Definition
 ) {
-    // natively, the attributes observed by the registered definition are going to be taken
+    const { observedAttributes, attributeChangedCallback } = instanceDefinition;
+    if (observedAttributes.size === 0 || isUndefined(attributeChangedCallback)) {
+        // This instance does not need to observe any attributes, no need to patch
+        return EMPTY_SET as Set<string>;
+    }
+
+    // Natively, the attributes observed by the registered definition are going to be taken
     // care of by the browser, only the difference between the two sets has to be taken
     // care by the patched version.
+    // TODO [#2877]: LWC components don't actually use observedAttributes, the size is never 0
```

Comment:
"...always 0"?

Reply:
Actually, this comment is wrong. Somehow I managed to get the code coverage to 100%; sometimes it is 0 and sometimes it's not. I can remove the comment.